keps/0009-operator-toolkit.md

```diff
@@ -0,0 +1,256 @@
+
+* **version**: String defining the version of a given operator
+* **tasks**: A map of tasks that can be run. These are the atomic runnable unit of a KUDO operator, and are made up of a series of YAML manifests. These are defined more in detail in [Tasks](#tasks).
+* **plans**: A map of plans that can be run. These are the core partitioning unit of a KUDO operator. A plan is intended to run a single "operations task" for an operator, such as backup, restore, deployment, or update. This is defined in detail in [Plans](#plans)
+
+This file undergoes a Go template pass on Instance instantiation before being parsed. This is described more in detail in [Extensions and Bases](#extensions-and-bases)
+
+#### params.yaml
+
+The `params.yaml` file is a list of keys and values for operator parameters and their defaults. In the MySQL example, this looks like:
+
+```yaml
+backupFile: backup.sql
+password: password
+```
+
+These values are meant to be overridden by Instance resources when instantiating an operator.
+
+This file undergoes a Go template pass on Instance instantiation before being parsed. This is described more in detail in [Parameters](#parameters).
+
+#### templates/
+
+The templates directory contains 0-to-many YAML manifests for operator tasks to use. These are described in more detail in [Templates](#templates).
+
+Each template undergoes a Go template pass for an Instance when it's corresponding Task is run.
+
+### Plans
+
+Plans are the core unit of operation within a KUDO operator. Each KUDO plan represents an individual unit of operations work. This may include backups, restores, deployments, upgrades, compaction, or any other myriad of operations that an application operator may want to perform with KUDO.
+
+A single plan is composed of [Steps](#steps).
+
+### Steps
+
+Steps are used for fine grained parition within a plan. Steps are represented as a list of objects that define the step name and a list of tasks to be run during that step.
+
+All tasks within the same step are applied at the same time. For runnable templates within a task (such as a Deployment, StatefulSet, Pod, etc), parallelism of workloads should be controlled through other Kubernetes primitives such as PodDisruptionBudget, and added to the task where the relevant workload is run.
+
+If a step contains a runnable task that has readiness probes defined, the step waits until the readiness probe succeeds before moving onto the next step.
+
+### Tasks
+
+Tasks are a map of task name to a list of templates that should be executed in that step. With [Extensions and Bases](#extensions-and-bases), tasks can be represented in multiple forms, as long as the end result is a ready-to-run YAML manifest for a Kubernetes object.
+
+#### Files
+If a filename is specified, KUDO will execute a Go Template on the relevant filename, described more in detail in [Templates](#templates).
+
+#### Webhooks
```

Comment:
I wouldn't be opposed to doing webhooks in phase 2 - it feels to me more like an extension than beta feature but if we feel like it's easy to implement, I am not opposed to having this in

Reply:
I would agree with this.  I'd rather focus on the controller stability before trying to fuse multiple input sources.

Reply:
Works for me. They're additive to the design and don't block a base implementation.

Reply:
@runyontr @alenkacz should I pull this out of the KEP or should we leave it as is but not do the stories for it at this time?

Reply:
I think it would be good to leave a comment about "There could/will be other rendering mechanisms, such as webhooks/cue, as long as they adhere to a conformant interface to be defined later".  Leave the idea/plan in there, but lets not spec out what that swagger API is yet.

Reply:
yep, agreed 👍 . Or we can put that into some section of possible future work as papers usually do that :) I think we will need a KEP for this, but we probably don't have to start one right now